engine/script/src/script_vmath.cpp

```diff
@@ -2299,6 +2401,293 @@ namespace dmScript
+	 * local v3 = vmath.vector4(30, -30, 45, 1)
+	 * print(vmath.clamp(v3, 0, 20)) -> vmath.vector4(20, 0, 20, 1)
+     * 
+     * local min_v = vmath.vector4(0, -10, -10, 1)
+     * print(vmath.clamp(v3, min_v, 20)) -> vmath.vector4(20, -10, 20, 1)
+     * ```
+     */
+    static int Vector_Clamp(lua_State* L)
+    {
+        if (lua_type(L, 1) == LUA_TNUMBER)
+        {
+            float value = (float) luaL_checknumber(L, 1);
+            float min = (float) luaL_checknumber(L, 2);
+            float max = (float) luaL_checknumber(L, 3);
+            lua_pushnumber(L, dmMath::Clamp(value, min, max));
+            return 1;
+        }
+
+        void* argument = 0;
+        const ScriptUserType argument_type = CheckUserData(L, 1, &argument);
+        int argument2_type = lua_type(L, 2);
+        int argument3_type = lua_type(L, 3);
+        if (argument_type == SCRIPT_TYPE_VECTOR3)
+        {
+            Vector3* value = (Vector3*)argument;
+            Vector3 min_v, max_v;
+            if (argument2_type == LUA_TNUMBER)
+            {
+                float min = (float) luaL_checknumber(L, 2);
+                min_v = Vector3(min, min, min);
+            }
+            else
+            {
+                min_v = *CheckVector3(L, 2);
+            }
+            if (argument3_type == LUA_TNUMBER)
+            {
+                float max = (float) luaL_checknumber(L, 3);
+                max_v = Vector3(max, max, max);
+            }
+            else
+            {
+                max_v = *CheckVector3(L, 3);
+            }
+
+            PushVector3(L, Vector3(
+                dmMath::Clamp(value->getX(), min_v.getX(), max_v.getX()),
+                dmMath::Clamp(value->getY(), min_v.getY(), max_v.getY()),
+                dmMath::Clamp(value->getZ(), min_v.getZ(), max_v.getZ())
+            ));
+        }
+        else if (argument_type == SCRIPT_TYPE_VECTOR4)
+        {
+            Vector4* value = (Vector4*)argument;
+            Vector4 min_v, max_v;
+            if (argument2_type == LUA_TNUMBER)
+            {
+                float min = (float) luaL_checknumber(L, 2);
+                min_v = Vector4(min, min, min, min);
+            }
+            else
+            {
+                min_v = *CheckVector4(L, 2);
+            }
+            if (argument3_type == LUA_TNUMBER)
+            {
+                float max = (float) luaL_checknumber(L, 3);
+                max_v = Vector4(max, max, max, max);
+            }
+            else
+            {
+                max_v = *CheckVector4(L, 3);
+            }
+            PushVector4(L, Vector4(
+                dmMath::Clamp(value->getX(), min_v.getX(), max_v.getX()),
+                dmMath::Clamp(value->getY(), min_v.getY(), max_v.getY()),
+                dmMath::Clamp(value->getZ(), min_v.getZ(), max_v.getZ()),
+                dmMath::Clamp(value->getW(), min_v.getW(), max_v.getW())
+            ));
+        }
+        else if  (argument_type == SCRIPT_TYPE_VECTOR)
```

Comment:
While I do like the implementation, I also wonder about the use case.
I rarely see anyone using the Vector type, and I think I'd personally just wait until someone needs it before adding it.

We use the mantra "every bit counts", in terms of keeping the engine size as small as possible. Sometimes the change it warranted, and sometime we want to wait until someone asks for it. I believe this case might be the latter.
I don't have an overly strong opinion here.

@britzl what do you think?

Reply:
I agree. The Vector type is very rarely used, and if it is used it is likely only to describe an animation curve. I think we can remove it and add a note in the documentation about which types are not supported.

Reply:
I've removed `clamp` implementation for `vector` type but left check for `SCRIPT_TYPE_VECTOR` in `CheckUserData` to cover all types defined in `ScriptUserType`.